source/Body.cpp

```diff
@@ -388,60 +819,150 @@ void Body::SetStep(int step) const
-		// final frame.
-		if(!repeat)
-			frame = min(frame, lastFrame);
-		else if(frame >= frames)
+		// Handle any frameRate changes in the animation.
+		if(anim.rampUpRate != 0.0)
+		{
+			if(frameRate >= Body::MIN_FRAME_RATE && frameRate <= anim.frameRate)
+				frameRate += anim.rampUpRate;
+		}
+		else
+			frameRate = anim.frameRate;
+		// For when it needs to be applied to transition.
+		delayed = 0.f;
+
+		// If repeating, wrap the frame index by the total cycle time.
+		if(anim.repeat)
+			frame = fmod(frame, cycle);
+
+		if(!anim.rewind)
+		{
+			// If not repeating, frame should never go higher than the index of the
+			// final frame.
+			if(!anim.repeat)
+			{
+				frame = min(frame, lastFrame);
+				framePercentage = (frame + 1) / frames;
+				if(framePercentage >= anim.indicatePercentage)
+					stateReady = anim.indicateReady;
+				else
+					stateReady = false;
+			}
+			else if(frame >= frames)
+			{
+				stateReady = false;
+				// If in the delay portion of the loop, set the frame to zero.
+				frame = 0.f;
+			}
+		}
+		else if(frame >= lastFrame)
 		{
-			// If we're in the delay portion of the loop, set the frame to 0.
-			frame = 0.f;
+			// In rewind mode, once you get to the last frame, count backwards.
+			// Regardless of whether we're repeating, if the frame count gets to
+			// be less than 0, clamp it to 0.
+			frame = max(0.f, lastFrame * 2.f - frame);
+			stateReady = false;
 		}
+		nextFrame = static_cast<int>(frame);
+		if(nextFrame != prevFrame)
+			randomFrame = static_cast<int>(static_cast<float>(Random::Real()) * frames);
+		else
+			randomFrame += frame - nextFrame;
 	}
-	else if(frame >= lastFrame)
+	else
 	{
-		// In rewind mode, once you get to the last frame, count backwards.
-		// Regardless of whether we're repeating, if the frame count gets to
-		// be less than 0, clamp it to 0.
-		frame = max(0.f, lastFrame * 2.f - frame);
+		// Override any delay if the ship wants to jump, become disabled, or land.
+		bool ignoreDelay = transitionState == Body::BodyState::JUMPING ||
+							transitionState == Body::BodyState::DISABLED ||
+							transitionState == Body::BodyState::LANDING ||
+							transitionState == Body::BodyState::TRIGGER;
+
+		if(delayed >= anim.transitionDelay || ignoreDelay)
+		{
+			if(anim.rampDownRate != 0.0)
+			{
+				if(frameRate >= Body::MIN_FRAME_RATE && frameRate <= anim.frameRate)
+					frameRate -= anim.rampDownRate;
+			}
+			else
+				frameRate = anim.frameRate;
+			// Handle transitions.
+			if(anim.transitionFinish && !anim.transitionRewind)
+			{
+				// Finish the ongoing state's animation, then transition.
+				frame = min(frame, lastFrame);
+
+				if(frame >= lastFrame)
+					stateTransitionCompleted = true;
+			}
+			else if(!anim.transitionFinish && anim.transitionRewind)
+			{
+				// Rewind the ongoing state's animation, then transition.
+				frame = max(0.f, rewindFrame * 2.f - frame);
+
+				if(frame == 0.f)
```

Comment:
```suggestion
				if(!frame)
```
just my personal preference